### PR TITLE
Update to monolog v2

### PR DESCRIPTION
**Pull request by snapshotpl**

### CODE REVIEW

**Comment by snapshotpl** [comment]
@abacaphiliac look at this. I did monolog v1 and v2 compatibility and also keep zf2. I think this pr doesn't have any BC so we can release it as 2.5 or 2.6!

**Comment by snapshotpl** [comment]
PHP 7.1 also saved!

File: composer.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
     "require": {
         "php": "^7.1",
         "container-interop/container-interop": "^1.0",
-        "monolog/monolog": "^1.0.2",
+        "monolog/monolog": "^1.3 || ^2.0",
```
Comment (abacaphiliac) [inline composer.json]: oh, very nice. i wasn't expecting this. please give me a couple of days to really dig in.

File: test/EnliteMonologTest/Service/FormatterMock.php [hunk 2]
```diff
@@ -1,46 +1,92 @@
 <?php
 
+// @codingStandardsIgnoreFile
```
Comment (abacaphiliac) [inline test/EnliteMonologTest/Service/FormatterMock.php]: what are these about?

Comment (snapshotpl) [reply test/EnliteMonologTest/Service/FormatterMock.php]: To keep compatibility I needed to create conditional implementations of mock. After that PHPCS throw this error https://travis-ci.org/enlitepro/enlite-monolog/jobs/641132116#L321

File: composer.json (hunk 1 above)
Comment (snapshotpl) [reply composer.json]: I needed to define monolog:^1.3 because this version is first where psr-3 was implemented. And this interface is now a bridge between v1 and v2 in public interface of monolog.

**Comment by snapshotpl** [comment]
@abacaphiliac I test this changes in my project and works like charm!

File: test/EnliteMonologTest/Service/FormatterMock.php (hunk 2 above)
Comment (abacaphiliac) [reply test/EnliteMonologTest/Service/FormatterMock.php]: i see. i suppose it will be short-lived anyway. we can drop `monolog:^1` support in the next major version of this package.

File: src/EnliteMonolog/Service/MonologServiceFactory.php [hunk 3]
```diff
@@ -130,8 +130,18 @@ public function createHandler($container, MonologOptions $options, $handler)
...
 
         if (isset($handler['formatter'])) {
-            $formatter = $this->createFormatter($container, $handler['formatter']);
-            $instance->setFormatter($formatter);
+            if ($instance instanceof FormattableHandlerInterface
+                || !defined('Monolog\Logger::API')
+                || \Monolog\Logger::API === 1) {
+                $formatter = $this->createFormatter($container, $handler['formatter']);
+                $instance->setFormatter($formatter);
+            } else {
```
Comment (abacaphiliac) [inline src/EnliteMonolog/Service/MonologServiceFactory.php]: it seems we're missing coverage over this branch. will you please add a test?

File: test/ZF3/phpunit.xml.dist [hunk 4]
```diff
@@ -6,6 +6,7 @@
          beStrictAboutOutputDuringTests="true"
          beStrictAboutTestsThatDoNotTestAnything="true"
          forceCoversAnnotation="true"
+         convertWarningsToExceptions="false"
```
Comment (abacaphiliac) [inline test/ZF3/phpunit.xml.dist]: why was this necessary?

File: src/EnliteMonolog/Service/MonologServiceFactory.php (hunk 3 above)
Comment (abacaphiliac) [reply src/EnliteMonolog/Service/MonologServiceFactory.php]: apologies, i now see that `testCannotCreateWithFormatterWithoutCorrectInterface` covers this condition for `php:^7.2`.

**Comment by abacaphiliac** [comment]
@snapshotpl you did a really nice job on this. thanks for all of the extra compatibility care. when i merge this, i plan to do a squash. do you have any concerns with that?

**Comment by snapshotpl** [comment]
Let's do it!

Niedz., 26 sty 2020, 01:35 użytkownik Timothy Younger <
notifications@github.com> napisał:

File: test/ZF3/phpunit.xml.dist (hunk 4 above)
Comment (snapshotpl) [reply test/ZF3/phpunit.xml.dist]: There are some warnings in lower dependency versions. Not so important, so I decided to not break test when appear

**Comment by abacaphiliac** [comment]
@snapshotpl great job, again. thanks so much! would you be willing to do a short-array syntax update to follow this?

**Comment by snapshotpl** [comment]
Short array syntax isn't valuable to end user so we can do it any time. The best what we can do now its release support for monolog 2

**Comment by abacaphiliac** [comment]
@snapshotpl you're right about that. i'll get that done now, and i'll create a 2.6 project with the leftovers.

**Comment by abacaphiliac** [comment]
`v2.5.0` has been released! thanks again, @snapshotpl, for all of your hard work on this one.

https://github.com/enlitepro/enlite-monolog/releases/tag/v2.5.0

**Comment by snapshotpl** [comment]
@abacaphiliac nice teamwork! :) Can you bump packagist https://packagist.org/packages/enlitepro/enlite-monolog ? Or even better setup push notification on packagist to update version immediately next time

**Comment by abacaphiliac** [comment]
@snapshotpl i manually updated in Packagist, so this latest release should be available now.

i can see that the Packagist service is set up in the settings of this repository, but for some reason the auto-update isn't triggered. [debug instructions on Packagist](https://packagist.org/about#how-to-update-packages) appear to be written for those who are maintainers of those projects on Packagist (this package is not in my package list). i am only a collaborator of this repository on GitHub, not a maintainer on Packagist.

@ftdebugger @Sysaninster would either of you feel comfortable adding me as a maintainer of this package on Packagist?
https://packagist.org/packages/enlitepro/enlite-monolog

**Comment by Sysaninster** [comment]
> @snapshotpl i manually updated in Packagist, so this latest release should be available now.
> 
> i can see that the Packagist service is set up in the settings of this repository, but for some reason the auto-update isn't triggered. [debug instructions on Packagist](https://packagist.org/about#how-to-update-packages) appear to be written for those who are maintainers of those projects on Packagist (this package is not in my package list). i am only a collaborator of this repository on GitHub, not a maintainer on Packagist.
> 
> @ftdebugger @Sysaninster would either of you feel comfortable adding me as a maintainer of this package on Packagist?
> https://packagist.org/packages/enlitepro/enlite-monolog

Unfortunately, I don't have the maintainer access to https://packagist.org/packages/enlitepro/enlite-monolog package on Packagist. But I sent the e-mail to @ftdebugger where I requested access to the package for you. I hope he didn't change his e-mail.

**Comment by abacaphiliac** [comment]
@Sysaninster thank you! i have access now. i'll debug auto-update this evening.

**Comment by abacaphiliac** [comment]
@ftdebugger thank you for adding me as a maintainer on Packagist. now that this package is a part of my Packagist profile, [triggering a manual account sync](https://packagist.org/about#how-to-update-packages) allowed Packagist to update the hooks on our behalf.

@snapshotpl auto-update has been restored.
